hudi-client/hudi-spark-client/src/main/scala/org/apache/spark/sql/hudi/HoodieSparkRecordSerializer.scala

```diff
@@ -97,25 +90,26 @@ class SparkStructTypeSerializer(schemas: Map[Long, StructType]) extends KSeriali
    */
   def serializeDatum(datum: HoodieSparkRecord, output: Output): Unit = {
     val schema = datum.getStructType
-    val fingerprint = fingerprintCache.getOrElseUpdate(schema, {
-      SchemaNormalization.fingerprint64(schema.json.getBytes(StandardCharsets.UTF_8))
-    })
-    schemas.get(fingerprint) match {
-      case Some(_) =>
-        output.writeBoolean(true)
-        output.writeLong(fingerprint)
-      case None =>
-        output.writeBoolean(false)
-        val compressedSchema = compress(schema)
-        output.writeInt(compressedSchema.length)
-        output.writeBytes(compressedSchema)
+    if (HoodieInternalRowUtils.containsCompressedSchema(schema)) {
+      val fingerprint = HoodieInternalRowUtils.getCachedFingerPrintFromSchema(schema)
+      output.writeBoolean(true)
+      output.writeLong(fingerprint)
+    } else {
+      output.writeBoolean(false)
+      val compressedSchema = compress(schema)
+      output.writeInt(compressedSchema.length)
+      output.writeBytes(compressedSchema)
     }
 
     val record = datum.newInstance().asInstanceOf[HoodieSparkRecord]
     record.setStructType(null)
-    val stream = new ObjectOutputStream(output)
+    val byteStream = new ByteArrayOutputStream()
```

Comment:
Sorry, i skimmed t/h this initially but overlooked how we're doing the serialization here:

 -   We should only serialize the `UnsafeRow` that HoodieSparkRecord holds (see my comment above); `UnsafeRow` implements KryoSerializable so we can use these methods to properly write it out into the stream; we should not be relying on Java's object serialization framework for that